liquigraph-core/src/main/java/org/liquigraph/core/io/lock/LockableConnection.java

```diff
@@ -0,0 +1,129 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.liquigraph.core.io.lock;
+
+import org.liquigraph.core.exception.LiquigraphLockException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.SQLException;
+import java.sql.Statement;
+import java.util.UUID;
+
+public final class LockableConnection implements ConnectionSupplier {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(LockableConnection.class);
+    private final Connection connection;
+    private final UUID uuid;
+    private final Thread task;
+
+    public LockableConnection(Connection connection) {
+        this.connection = connection;
+        this.uuid = UUID.randomUUID();
+        this.task = new Thread(new ShutdownTask());
+        initiateLocking();
+    }
+
+    private final void initiateLocking() {
+        registerLockRemovalHook();
+        ensureLockUnicity();
+        tryWriteLock();
+    }
+
+   /**
+    * Closes the underlying connection
+    */
+   @Override
+    public void close() {
+        Runtime.getRuntime().removeShutdownHook(task);
+        closeConnection();
+    }
+
+    @Override
+    public Connection get() {
+        return connection;
+    }
+
+    private final void registerLockRemovalHook() {
+        Runtime.getRuntime().addShutdownHook(task);
+    }
+
+    private final void ensureLockUnicity() {
+        try (Statement statement = connection.createStatement()) {
+            statement.execute("CREATE CONSTRAINT ON (lock:__LiquigraphLock) ASSERT lock.name IS UNIQUE");
+            connection.commit();
+        }
+        catch (SQLException e) {
+            throw new LiquigraphLockException(
+                "Could not ensure __LiquigraphLock unicity\n\t" +
+                    "Please make sure your instance is in a clean state\n\t" +
+                    "No more than 1 lock should be there simultaneously!",
+                e
+            );
+        }
+    }
+
+    private final void tryWriteLock() {
+        try (PreparedStatement statement = connection.prepareStatement(
+            "CREATE (:__LiquigraphLock {name:'John', uuid:{1}})")) {
+
+            statement.setString(1, uuid.toString());
+            statement.execute();
+            connection.commit();
+        }
+        catch (SQLException e) {
+            throw new LiquigraphLockException(
+                "Cannot create __LiquigraphLock lock\n\t" +
+                    "Likely another Liquigraph execution is going on or has crashed.",
+                e
+            );
+        }
+    }
+
+    private final void closeConnection() {
+        try {
+            if (!connection.isClosed()) {
+                task.run();
```

Comment:
When the connection is closed too early, either the lock is never removed (current impl) or removed too early (when extending Connection). I'm not sure what's best.

In any case, you are right, I should and will add more documentation.